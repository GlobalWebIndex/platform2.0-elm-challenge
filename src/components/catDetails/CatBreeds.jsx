import React from "react";
import PropTypes from "prop-types";
import { Link, useLocation } from "react-router-dom";

const CatBreeds = ({ breeds = [] }) => {
  const location = useLocation();

  if (!breeds?.length) {
    return null;
  }

  return (
    // Intentionally this is a div. We want a 'block-level' element.
    <div>
      <strong>Breeds: </strong>
      {breeds.map((breed) => (
        <Link
          to={`/breed/${breed.id}`}
          state={{ backgroundLocation: location }}
          key={breed.id}
        >
          {breed.name}
        </Link>
      ))}
    </div>
  );
};

CatBreeds.propTypes = {
  breeds: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
    })
  ),
};

export default CatBreeds;
