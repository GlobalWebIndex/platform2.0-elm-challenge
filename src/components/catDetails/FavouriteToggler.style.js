import { createUseStyles } from "react-jss";

export const useStyles = createUseStyles({
  favouriteToggler: {
    marginTop: 15,
  },
  favouriteButton: {
    backgroundColor: "#8b0000",
    color: "#fff",
    display: "block",
    textAlign: "center",
    textTransform: "uppercase",
    padding: [15, 60],
    margin: [0, "auto"],
    cursor: "pointer",
    outline: "none",
    border: 0,
    minWidth: 300,

    "&:hover": {
      backgroundColor: "#3b0000",
    },
    favouriteIcon: {
      color: "#fff",
    },
  },
});
