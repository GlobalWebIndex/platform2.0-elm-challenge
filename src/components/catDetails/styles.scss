.container {
  .horizontal {
    display: flex;
    gap: 1rem;

    &.withNoBreeds {
      justify-content: center;
    }

    .image {
      width: 50%;

      img {
        width: 100%;
        object-fit: cover;
      }
    }

    .breeds {
      width: 50%;
    }
  }

  .vertical {
    display: flex;
    flex-direction: column;
    gap: 1rem;

    .image {
      width: 100%;

      img {
        width: 100%;
        object-fit: cover;
      }
    }

    .breeds {
      width: 100%;
    }
  }
}
