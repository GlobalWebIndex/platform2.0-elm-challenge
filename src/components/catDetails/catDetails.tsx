import React from 'react';
import classnames from 'classnames';
import type {Cat} from 'cat-lover/types/cats';
import {BreedInfo} from './breedInfo';
import styles from './styles.scss';

type CatDetailsProps = {
  cat: Cat;
};

export const CatDetails = ({cat}: CatDetailsProps): JSX.Element => {
  const hasBreeds = Boolean(cat.breeds?.length);
  const isHorizontalLayout = cat.width / cat.height <= 1;

  return (
    <div className={styles.container}>
      {isHorizontalLayout ? (
        <div
          className={classnames(styles.horizontal, {
            [styles.withNoBreeds]: !hasBreeds
          })}
        >
          <div className={styles.image}>
            <img alt="Cat" src={cat.url} />
          </div>
          {hasBreeds && <BreedInfo cat={cat} className={styles.breeds} />}
        </div>
      ) : (
        <div className={styles.vertical}>
          <div className={styles.image}>
            <img alt="Cat" src={cat.url} />
          </div>
          {hasBreeds && <BreedInfo cat={cat} className={styles.breeds} />}
        </div>
      )}
    </div>
  );
};
