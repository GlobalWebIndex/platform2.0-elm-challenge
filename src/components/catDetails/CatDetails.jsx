import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import FavouriteToggler from "./FavouriteToggler";
import Loading from "../common/Loading";
import Failure from "../common/Failure";
import { fetchCatDetails } from "../../util/api.js";
import { AsyncStatus, useAsync } from "../../hooks/useAsync";
import Card from "../cards/Card";
import CatBreeds from "./CatBreeds";

const CatDetails = () => {
  const { catId } = useParams();
  const { data: fetchedCatImg, status, error, run } = useAsync();

  useEffect(() => {
    const callback = fetchCatDetails(catId);
    run(callback);
  }, [run, catId]);

  switch (status) {
    case AsyncStatus.IDLE:
    case AsyncStatus.PENDING:
      return <Loading />;

    case AsyncStatus.REJECTED:
      return <Failure error={error} />;

    case AsyncStatus.RESOLVED:
      return (
        <Card
          imageSrc={fetchedCatImg.url}
          imageWidth={fetchedCatImg.width}
          imageHeight={fetchedCatImg.height}
          id={fetchedCatImg.id}
          title={fetchedCatImg.title}
        >
          <CatBreeds breeds={fetchedCatImg.breeds} />
          <FavouriteToggler imageId={fetchedCatImg.id} />
        </Card>
      );
    default:
      throw new Error("This shouldn't ever happen");
  }
};

CatDetails.propTypes = {};

export default CatDetails;
