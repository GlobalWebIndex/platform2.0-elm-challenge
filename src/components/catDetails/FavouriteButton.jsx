import React from "react";
import PropTypes from "prop-types";
import { useStyles } from "./FavouriteToggler.style";
import noop from "../../util/noop";

const FavouriteButton = ({ children, disabled = false, onClick = noop }) => {
  const classes = useStyles();

  return (
    <div className={classes.favouriteToggler}>
      <button
        className={classes.favouriteButton}
        disabled={disabled}
        onClick={onClick}
        data-testid="favourite-button"
      >
        {children}
      </button>
    </div>
  );
};

FavouriteButton.propTypes = {
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

export default FavouriteButton;
