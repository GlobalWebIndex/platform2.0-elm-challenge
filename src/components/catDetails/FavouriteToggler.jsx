import React, { useEffect, useState, useCallback } from "react";
import { HeartOutlined, HeartFilled, LoadingOutlined } from "@ant-design/icons";
import { addFavourite, fetchFavourites, removeFavourite } from "../../util/api";
import FavouriteButton from "./FavouriteButton";
import { useStyles } from "./FavouriteToggler.style";

const FavouriteToggler = ({ imageId = "" }) => {
  const classes = useStyles();
  const [favourite, setFavourite] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const fetchAsyncFavourite = useCallback(async () => {
    setIsLoading(true);
    const response = await fetchFavourites();

    if (response?.length) {
      const foundFavourite =
        response.find((favourite) => favourite.image_id === imageId) || null;
      setFavourite(foundFavourite);
    }
    setIsLoading(false);
  }, [imageId]);

  const onToggleFavourite = useCallback(async () => {
    setIsLoading(true);

    // TODO: Error handling could be definitely improved here!
    if (favourite) {
      await removeFavourite(favourite.id);
      setFavourite(null);
    } else {
      const response = await addFavourite(imageId);
      setFavourite(response);
    }

    setIsLoading(false);
  }, [imageId, favourite]);

  useEffect(() => {
    fetchAsyncFavourite();
  }, [fetchAsyncFavourite]);

  if (!imageId) {
    return null;
  }

  if (isLoading) {
    return (
      <FavouriteButton disabled>
        <LoadingOutlined />
      </FavouriteButton>
    );
  }

  if (favourite) {
    return (
      <FavouriteButton onClick={onToggleFavourite}>
        <HeartFilled className={classes.favouriteIcon} /> Remove from favourites
      </FavouriteButton>
    );
  }

  return (
    <FavouriteButton onClick={onToggleFavourite}>
      <HeartOutlined className={classes.favouriteIcon} /> Add as a favourite!
    </FavouriteButton>
  );
};

FavouriteToggler.propTypes = {};

export default FavouriteToggler;
