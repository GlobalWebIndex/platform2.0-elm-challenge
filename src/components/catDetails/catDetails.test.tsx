import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render} from '@testing-library/react';
import type {Cat} from 'cat-lover/types/cats';
import {CatDetails} from './catDetails';

describe('<CatDetails />', () => {
  describe('when the image width is less than or equal to its height', () => {
    it('should render with the horizontal layout', () => {
      const mockCat: Cat = {
        breeds: [],
        height: 500,
        id: '',
        url: 'https://example.com/cat.jpg',
        width: 400
      };

      const {container} = render(
        <MemoryRouter initialEntries={['/cats/1']}>
          <CatDetails cat={mockCat} />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });

    describe('when breeds are available', () => {
      it('should render breed information', () => {
        const mockCat: Cat = {
          breeds: [
            {
              country_code: '',
              country_codes: '',
              id: '1',
              life_span: '10-12',
              name: 'Breed 1',
              origin: 'Origin 1',
              temperament: 'Temperament 1',
              weight: {
                imperial: '',
                metric: ''
              },
              wikipedia_url: 'http://example.com/breed1'
            }
          ],
          height: 500,
          id: '',
          url: 'http://example.com/cat.jpg',
          width: 400
        };

        const {container} = render(
          <MemoryRouter initialEntries={['/cats/1']}>
            <CatDetails cat={mockCat} />
          </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
      });
    });
  });

  describe('when the image width is greater than the height', () => {
    it('should render with the vertical layout', () => {
      const mockCat: Cat = {
        breeds: [],
        height: 400,
        id: '',
        url: 'http://example.com/cat.jpg',
        width: 500
      };

      const {container} = render(
        <MemoryRouter initialEntries={['/cats/1']}>
          <CatDetails cat={mockCat} />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });

    describe('when breeds are available', () => {
      it('should render breed information', () => {
        const mockCat: Cat = {
          breeds: [
            {
              country_code: '',
              country_codes: '',
              id: '1',
              life_span: '10-12',
              name: 'Breed 1',
              origin: 'Origin 1',
              temperament: 'Temperament 1',
              weight: {
                imperial: '',
                metric: ''
              },
              wikipedia_url: 'http://example.com/breed1'
            }
          ],
          height: 400,
          id: '',
          url: 'http://example.com/cat.jpg',
          width: 500
        };

        const {container} = render(
          <MemoryRouter initialEntries={['/cats/1']}>
            <CatDetails cat={mockCat} />
          </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
      });
    });
  });
});
