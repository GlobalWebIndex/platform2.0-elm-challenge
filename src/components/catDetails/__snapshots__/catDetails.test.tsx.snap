// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatDetails /> when the image width is greater than the height should render with the vertical layout 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="vertical"
    >
      <div
        class="image"
      >
        <img
          alt="Cat"
          src="http://example.com/cat.jpg"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<CatDetails /> when the image width is greater than the height when breeds are available should render breed information 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="vertical"
    >
      <div
        class="image"
      >
        <img
          alt="Cat"
          src="http://example.com/cat.jpg"
        />
      </div>
      <div
        class="breeds"
      >
        <h2>
          Breeds
        </h2>
        <ul>
          <li>
            <h3>
              Breed 1
            </h3>
            <ul>
              <li>
                <strong>
                  Origin:
                </strong>
                 
                Origin 1
              </li>
              <li>
                <strong>
                  Temperament:
                </strong>
                 
                Temperament 1
              </li>
              <li>
                <strong>
                  Life span:
                </strong>
                 
                10-12
                 years
              </li>
              <li>
                <strong>
                  Wikipedia:
                </strong>
                 
                <a
                  href="http://example.com/breed1"
                  rel="noreferrer"
                  target="_blank"
                >
                  http://example.com/breed1
                </a>
              </li>
              <li>
                <a
                  href="/breeds/1"
                >
                  Visit the breed page
                </a>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CatDetails /> when the image width is less than or equal to its height should render with the horizontal layout 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="horizontal withNoBreeds"
    >
      <div
        class="image"
      >
        <img
          alt="Cat"
          src="https://example.com/cat.jpg"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<CatDetails /> when the image width is less than or equal to its height when breeds are available should render breed information 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="horizontal"
    >
      <div
        class="image"
      >
        <img
          alt="Cat"
          src="http://example.com/cat.jpg"
        />
      </div>
      <div
        class="breeds"
      >
        <h2>
          Breeds
        </h2>
        <ul>
          <li>
            <h3>
              Breed 1
            </h3>
            <ul>
              <li>
                <strong>
                  Origin:
                </strong>
                 
                Origin 1
              </li>
              <li>
                <strong>
                  Temperament:
                </strong>
                 
                Temperament 1
              </li>
              <li>
                <strong>
                  Life span:
                </strong>
                 
                10-12
                 years
              </li>
              <li>
                <strong>
                  Wikipedia:
                </strong>
                 
                <a
                  href="http://example.com/breed1"
                  rel="noreferrer"
                  target="_blank"
                >
                  http://example.com/breed1
                </a>
              </li>
              <li>
                <a
                  href="/breeds/1"
                >
                  Visit the breed page
                </a>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
